README.md
moov-io/paygate
===

[![GoDoc](https://godoc.org/github.com/moov-io/paygate?status.svg)](https://godoc.org/github.com/moov-io/paygate)
[![Build Status](https://travis-ci.com/moov-io/paygate.svg?branch=master)](https://travis-ci.com/moov-io/paygate)
[![Coverage Status](https://codecov.io/gh/moov-io/paygate/branch/master/graph/badge.svg)](https://codecov.io/gh/moov-io/paygate)
[![Go Report Card](https://goreportcard.com/badge/github.com/moov-io/paygate)](https://goreportcard.com/report/github.com/moov-io/paygate)
[![Apache 2 licensed](https://img.shields.io/badge/license-Apache2-blue.svg)](https://raw.githubusercontent.com/moov-io/paygate/master/LICENSE)

*project is under active development and is not production ready*

Moov Paygate is a RESTful API enabling Automated Clearing House ([ACH](https://en.wikipedia.org/wiki/Automated_Clearing_House)) transactions to be submitted and received without a deep understanding of a full NACHA file specification.

Docs: [docs.moov.io](https://docs.moov.io/paygate/) | [api docs](https://api.moov.io/apps/paygate/)

## Project Status

This project is currently pre-production and could change without much notice, however we are looking for community feedback so please try out our code or give us feedback!

## Getting Started

Paygate can be ran or deployed in various ways. We have several guides for running paygate and offer a testing utility called [`apitest` from the moov-io/api repository](https://github.com/moov-io/api#apitest) for verifying paygate (and its dependnecies) are running properly.

- [Using docker-compose](#local-development)
- [Using our Docker image](#docker-image)
- [Build from source](#build-from-source)
- [How to setup open source ACH payments using Moov.io suite](https://medium.com/@tgunnoe/how-to-setup-open-source-ach-payments-using-moov-io-suite-3586757e45d6) by Taylor Gunnoe
  - Taylor has also written [paygate-cli](https://github.com/tgunnoe/paygate-cli) which is a command-line interface to paygate.

### Vocab

<dl>
    <dt>CIP</dt><dd>Customer Identification Program -- Verification process of customer information including KYC and social security number</dd>
    <dt>customerID</dt><dd>Unique ID from the <a href="https://github.com/moov-io/customers">Moov Customers</a> service. KYC and CIP checks utilize this to associate with paygate objects.</dd>
    <dt>KYC</dt><dd>Know Your Customer -- Verification process of customer information including: name, date of birth and address</dd>
    <dt>Originator</dt><dd>Entity that initiates the ACH transfer. Often usecases have one Originator to make all transfers from.</dd>
    <dt>Receiver</dt><dd>Entity that is the target of an ACH transfer.</dd>
    <dt>userID</dt><dd>Unique ID used to organize and isolate objects created within paygate. Objects are not accessible across userIDs</dd>
</dl>

## Deployment

Paygate currently requires the following services to be deployed and available:

- [ACH](https://github.com/moov-io/ach) (HTTP Server) via `ACH_ENDPOINT`
- [FED](https://github.com/moov-io/fed)  (HTTP Server) via `FED_ENDPOINT`
- [OFAC](https://github.com/moov-io/ofac) (HTTP Server) via `OFAC_ENDPOINT`
- The `X-User-Id` (case insensntive) HTTP header is also required and we recommend using an auth proxy to set this. Paygate only expects this value to be unique and consistent to a user.

The following services are required by default, but can be disabled:

- [Accounts](https://github.com/moov-io/accounts) (HTTP server) via `ACCOUNTS_ENDPOINT` and disabled with `ACCOUNTS_CALLS_DISABLED=yes`

### Docker image

You can download [our docker image `moov/paygate`](https://hub.docker.com/r/moov/paygate/) from Docker Hub or use this repository. No configuration is required to serve on `:8082` and metrics at `:9092/metrics` in Prometheus format.


```
$ docker run -p 8082:8082 moov/paygate:latest
ts=2018-12-13T19:18:11.970293Z caller=main.go:55 startup="Starting paygate server version v0.5.1"
ts=2018-12-13T19:18:11.970391Z caller=main.go:59 main="sqlite version 3.25.2"
ts=2018-12-13T19:18:11.971777Z caller=database.go:88 sqlite="starting database migrations"
ts=2018-12-13T19:18:11.971886Z caller=database.go:97 sqlite="migration #0 [create table if not exists receivers(cus...] changed 0 rows"
... (more database migration log lines)
ts=2018-12-13T19:18:11.97221Z caller=database.go:100 sqlite="finished migrations"
ts=2018-12-13T19:18:11.974316Z caller=main.go:96 ach="Pong successful to ACH service"
ts=2018-12-13T19:18:11.975093Z caller=main.go:155 transport=HTTP addr=:8082
ts=2018-12-13T19:18:11.975177Z caller=main.go:124 admin="listening on :9092"

$ curl -XPOST -H "x-user-id: test" localhost:8082/originators --data '{...}'
```

### Local development

We support a [Docker Compose](https://docs.docker.com/compose/gettingstarted/) environment in paygate that can be used to launch the entire Moov stack. After setup launching the stack is the following steps and we offer a testing utility [`apitest` from the moov-io/api repository](https://github.com/moov-io/api#apitest).

```
$ docker-compose up -d
paygate_ach_1 is up-to-date
paygate_ofac_1 is up-to-date
Recreating paygate_accounts_1 ...
paygate_fed_1 is up-to-date
Recreating paygate_accounts_1 ... done
Recreating paygate_paygate_1  ... done

# Run Moov's testing utility
$ apitest -local
2019/06/10 21:18:06.117261 main.go:61: Starting apitest v0.9.5
2019/06/10 21:18:06.117293 main.go:133: Using http://localhost as base API address
...
2019/06/10 21:18:06.276443 main.go:218: SUCCESS: Created user b1f2671bbed52ed6da88f16ce467cadecb0ee1b6 (email: festive.curran27@example.com)
...
2019/06/10 21:18:06.607817 main.go:218: SUCCESS: Created USD 204.71 transfer (id=b7ecb109574187ff726ba48275dcf88956c26841) for user
```

### Build from source

PayGate orchestrates several services that depend on Docker and additional GoLang libraries to run. Paygate leverages [Go Modules](https://github.com/golang/go/wiki/Modules) to manage dependencies. Ensure that your build environment is running Go 1.11 or greater and the environment variable `export GO111MODULE=on` is set. PayGate depends on other Docker containers that will be downloaded for testing and running the service. Ensure [Docker](https://docs.docker.com/get-started/) is installed and running.

```
$ cd moov/paygate # wherever this project lives

$ go run .
ts=2018-12-13T19:18:11.970293Z caller=main.go:55 startup="Starting paygate server version v0.5.1"
ts=2018-12-13T19:18:11.970391Z caller=main.go:59 main="sqlite version 3.25.2"
ts=2018-12-13T19:18:11.971777Z caller=database.go:88 sqlite="starting database migrations"
ts=2018-12-13T19:18:11.971886Z caller=database.go:97 sqlite="migration #0 [create table if not exists receivers(cus...] changed 0 rows"
... (more database migration log lines)
ts=2018-12-13T19:18:11.97221Z caller=database.go:100 sqlite="finished migrations"
ts=2018-12-13T19:18:11.974316Z caller=main.go:96 ach="Pong successful to ACH service"
ts=2018-12-13T19:18:11.975093Z caller=main.go:155 transport=HTTP addr=:8082
ts=2018-12-13T19:18:11.975177Z caller=main.go:124 admin="listening on :9092"
```

### Configuration

Paygate supports reading a config file or environmental variables. Below is a sample config file:

```yaml
log_format: json

ach:
  storage_dir: /data/paygate/ach/

sqlite:
  path: /data/paygate/paygate.db

odfi:
  account_number: 214115514
  account_type: checking
  bank_name: Moov
  holder: Jane Smith
  identification: 29813754
  routing_number: 987654320

web:
  bind_address: 0.0.0.0:8080
```

The file can contain both [routing information](testdata/configs/routing-good.yaml) and [general configuration](testdata/configs/valid.yaml) in one file.

#### Environment Variables

The following environmental variables can be set to configure behavior in paygate.

| Environmental Variable | Description | Default |
|-----|-----|-----|
| `ACH_ENDPOINT` | DNS record responsible for routing us to an [ACH](https://github.com/moov-io/ach) instance. If running as part of our local development setup (or in a Kubernetes cluster we setup) you won't need to set this. | `http://ach.apps.svc.cluster.local:8080/` |
| `ACCOUNTS_ENDPOINT` | A DNS record responsible for routing us to an [Accounts](https://github.com/moov-io/accounts) instance. | `http://accounts.apps.svc.cluster.local:8080` |
| `ACCOUNTS_CALLS_DISABLED=yes` | Flag to completely disable all calls to an Accounts service. This is used when paygate doesn't need to integrate with a general ledger solution. | `no` |
| `FED_ENDPOINT` | HTTP address for [FED](https://github.com/moov-io/fed) interaction to lookup ABA routing numbers. | `http://fed.apps.svc.cluster.local:8080` |
| `HTTP_ADMIN_BIND_ADDRESS` | Address for paygate to bind its admin HTTP server on. This overrides the command-line flag `-admin.addr`. | `:9092` |
| `HTTP_BIND_ADDRESS` | Address for paygate to bind its HTTP server on. This overrides the command-line flag `-http.addr`. | `:8082` |
| `HTTP_CLIENT_CAFILE` | Filepath for additional (CA) certificates to be added into each `http.Client` used within paygate. | Empty |
| `HTTPS_CERT_FILE` | Filepath containing a certificate (or intermediate chain) to be served by the HTTP server. Requires all traffic be over secure HTTP. | Empty |
| `HTTPS_KEY_FILE`  | Filepath of a private key matching the leaf certificate from `HTTPS_CERT_FILE`. | Empty |
| `LOG_FORMAT` | Format for logging lines to be written as. (Options: `json`, `plain`) | `plain` |
| `OFAC_ENDPOINT` | HTTP address for [OFAC](https://github.com/moov-io/ofac) interaction, defaults to Kubernetes inside clusters and local dev otherwise. | `http://ofac.apps.svc.cluster.local:8080` |
| `OFAC_MATCH_THRESHOLD` | Percent match against OFAC data that's required for paygate to block a transaction. | `0.99` |
| `DATABASE_TYPE` | Which database option to use - See **Storage** header below for per-database configuration (Options: `sqlite`, `mysql`) | `sqlite` |
| `CONFIG_FILEPATH` | File path if given will load configs from a Yaml file instead of a database. | Empty |


#### ACH file uploading / transfers

| Environmental Variable | Description | Default |
|-----|-----|-----|
| `ACH_FILE_BATCH_SIZE` | Number of Transfers to retrieve from the database in each batch for mergin before upload to Fed. | 100 |
| `ACH_FILE_MAX_LINES` | Maximum line count before an ACH file is uploaded to its remote server. NACHA guidelines have a hard limit of 10,000 lines. | 10000 |
| `ACH_FILE_TRANSFERS_CAFILE` | Filepath for additional (CA) certificates to be added into each FTP client used within paygate. | Empty |
| `ACH_FILE_TRANSFER_INTERVAL` | Go duration for how often to check and sync ACH files on their SFTP destinations. (Set to `0m` to disable.) | `10m` |
| `ACH_FILE_STORAGE_DIR` | Filepath for temporary storage of ACH files. This is used as a scratch directory to manage outbound and incoming/returned ACH files. | `./storage/` |
| `FORCED_CUTOFF_UPLOAD_DELTA` | Go duration for when the current time is within the routing number's cutoff time by duration force that file to be uploaded. | `5m` |

See [our detailed documentation for FTP and SFTP configurations](docs/ach.md#uploads-of-merged-ach-files).

##### FTP Configuration

Our FTP client offers some configuration options. Paygate currently uses the [jlaffaye/ftp](https://github.com/jlaffaye/ftp) library.

| Environmental Variable | Description | Default |
|-----|-----|-----|
| `FTP_DIAL_TIMEOUT` | Go duration for timeout when creating FTP connections. | `10s` |
| `FTP_DIAL_WITH_DISABLED_ESPV` | Offer EPSV to be used if the FTP server supports it. | `false` |

##### SFTP Configuration

Our SFTP (SSH File Transfer Protocol) client offers some configuration options. Paygate currently uses the [pkg/sftp](https://github.com/pkg/sftp) library.

| Environmental Variable | Description | Default |
|-----|-----|-----|
| `SFTP_DIAL_TIMEOUT` | Go duration for timeout when creating SFTP connections. | `10s` |
| `SFTP_MAX_CONNS_PER_FILE` | Sets the maximum concurrent requests allowed for a single file. | 8 |
| `SFTP_MAX_PACKET_SIZE` | Sets the maximum size of the payload, measured in bytes. Try lowering this on "failed to send packet header: EOF" errors. | 20480 |

Note: By default paygate **does not verify** the SFTP host public key. Write the expected public key into `sftp_configs`'s `host_public_key` column to have paygate verify.

#### Micro Deposits

In order to validate `Depositories` and transfer money paygate must submit small deposits and credits and have someone confirm the amounts manually. This is only required once per `Depository`. The configuration options for paygate are below and are all required:

| Environmental Variable | Description | Default |
|-----|-----|-----|
| `ODFI_ACCOUNT_NUMBER` | Account Number of Financial Institution which is originating micro deposits. | 123 |
| `ODFI_ACCOUNT_TYPE` | Type of ODFI account used for micro-depsits (Checking or Savings) | Savings |
| `ODFI_BANK_NAME` | Legal name of Financial Institution which is originating micro deposits. | Moov, Inc |
| `ODFI_HOLDER` | Legal name of Financial Institution which is originating micro deposits. | Moov, Inc |
| `ODFI_IDENTIFICATION` | Number by which the customer is known to the Financial Institution originating micro deposits. | 001 |
| `ODFI_ROUTING_NUMBER` | ABA routing number of Financial Institution which is originating micro deposits. | 121042882 |

#### Storage

Based on `DATABASE_TYPE` the following environment variables will be read to configure connections for a specific database.

##### MySQL

| Environmental Variable | Description | Default |
|-----|-----|-----|
| `MYSQL_HOSTNAME` | Hostname for connecting to the mysql server. (Example `mysql`) | Empty |
| `MYSQL_PORT` | Port for connecting to the mysql server. | `3306` |
| `MYSQL_PROTOCOL` | Protocol for connecting to the mysql server. (Example `tcp | unix`) | `tcp` |
| `MYSQL_DATABASE` | Name of database to connect into. | Empty |
| `MYSQL_PASSWORD` | Password of user account for authentication. | Empty |
| `MYSQL_USER` | Username used for authentication. | Empty |
| `MYSQL_TIMEOUT` | Timeout value to use when makeing calls to mysql. | `30s` |
| `MYSQL_STARTUP_TIMEOUT` | Timeout value to wait for port to be open when connecting to mysql. | Empty |

Refer to the mysql driver documentation for [connection parameters](https://github.com/go-sql-driver/mysql#dsn-data-source-name).

##### SQLite

| Environmental Variable | Description | Default |
|-----|-----|-----|
| `SQLITE_DB_PATH` | Local filepath location for the paygate SQLite database. | `paygate.db` |

Refer to the sqlite driver documentation for [connection parameters](https://github.com/mattn/go-sqlite3#connection-string).

## Getting Help

 channel | info
 ------- | -------
 [Project Documentation](https://docs.moov.io/) | Our project documentation available online.

CHANGELOG.md
## v0.8.0 (Unreleased)

Version v0.8.0 of PayGate adds some Prometheus metrics, basic transfer limits, and an OpenAPI specification for admin endpoints. Along with internal refactoring to cleanup code we have removed the `Depository` account number migration from previous releases.

`Transfer` objects have an additional status called `reviewable` which prevents them from being uploaded to an ODFI and requires their manual approval with admin endpoints.

ADDITIONS

- admin: Include generated Go client code and OpenAPI specification
- filetransfer: add ach_file_upload_errors for tracking ACH upload errors
- transfers: introduce basic calculations for N-day transfer limits
- transfers: store the client's real ip address on creation
- transfers: set status to 'reviewable' if over the limit
- transfers: add admin route for selective transfer status updates
- filetransfer: remove micro-deposits from mergable file when the Depository is deleted
- filetransfer: add admin endpoint for merging outgoing files

IMPROVEMENTS

- filetransfer: disable `Depository` field updates from NOC's, allow via config (`UPDATE_DEPOSITORIES_FROM_CHANGE_CODE=yes`)
- filetransfer: reject related objects from COR/NOC when not auto-updating fields
- filetransfer: rename admin HTTP routes to /configs/filetransfers/*
- api: use shared Error model
- api,client: clarify Address and BirthDate for KYC is not persisted
- verification/microdeposit: move routes, repository, etc into subpackage
- filetransfer: remove batches when deleting non-uploaded transfer
- transfers: refactor Receiver status check back into transfers package

BUG FIXES

- admin: fix micro-deposit return unmarshal
- filetransfer: fix partial updating of FileTransferConfig in admin HTTP routes
- filetransfer: handle nil FTPTransferAgent in Close
- depository: in admin update status route read userID from path
- originators,receivers: return null BirthDate,Address if empty
- filetransfer,transfers: update status once Transfers are merged, not in MarkTransferAsMerged
- transfers: fix amount storage truncation

BUILD

- build: run sonatype-nexus-community/nancy in CI
- build: upgrade Watchman to v0.13.2
- client: generate with 'client' as pacakge name
- docker: Update golang Docker tag to v1.14

## v0.7.3 (Released 2020-03-16)

BUG FIXES

- config: read "yes" or "true" properly

## v0.7.2 (Released 2020-03-16)

BUILD

- build: upload `paygate.exe` on each release

## v0.7.1 (Released 2020-01-22)

BUG FIXES

- api,client: There was a mistaken character in the OpenAPI docs `Ç` which should have been `C`.

IMPROVEMENTS

- chore(deps): update moov/ach docker tag to v1.3.0

BUILD

- build: add slack notifications
- build: upgrade golangci-lint to v1.23.1

## v0.7.0 (Released 2020-01-15)

Version v0.7.0 of Paygate adds more support for return and incoming (COR/NOC) file processing along with admin methods for updating file transfer configuration without needing to restart paygate or update the underlying database directly.

Account numbers are encrypted in a migration as part of this release. Masked versions are stored alongside their encrypted form, but the cleartext values are not wiped.

Please [read over the guide for account number encryption migration](docs/account-number-encryption.md#encrypted-account-numbers).

Also included are [filename templates](https://docs.moov.io/paygate/ach/#filename-templates) for merged ACH files uploaded to an ODFI. This allows banks with specific rules for files to be used with paygate.

KYC information is now optionally read for `Originator` and `Receiver` objects on creation. Paygate uses Moov Customers to perform validation and an OFAC check of this data.

The Moov Customers service will be used to register each Originator and Receiver (and their Depository) objects with any disclaimers required to be accepted prior to Transfer creation. OFAC searches for each customer are refreshed weekly by default with a config option to change the allowed staleness.

ADDITIONS

- internal/filetransfer: add HTTP routes for updating and deleting file transfer configs
- internal/filetransfer: allow synchronous waiting for flush routes (`/files/flush?wait`)
- main: Add 'GET /version' admin endpoint
- api,client: add return codes to Depository and Transfer HTTP responses
- filetransfer: support reading a config file for routing and FTP/SFTP configuration
- filetransfer: update Depository and Originator/Receiver objects from incoming COR entries
- depositories: add admin route for overriding status
- micro-deposits: record metrics on initate and confirmation
- micro-deposits: prevent additional attempts once we've failed too many times
- customers: Refresh OFAC searches weekly by default
- filetransfer: reject file uploads if they're not whitelisted by IP address

IMPROVEMENTS

- micro-deposits: don't require x-user-id on admin route to read
- internal/filetransfer: override String() on ftp and sftp configs to hide passwords
- micro-deposits: debit the micro-deposit from the remote account when crediting
- micro-deposits: accumulate deposit amounts for a final withdraw
- files: reverse micro-deposit transactions in Accounts, if enabled
- internal/database: micro_deposits.return_code should default to an empty string
- internal/filetransfer: support additional NACHA return codes
- internal/filetransfer: pass through x-request-id and x-user-id HTTP headers
- all: wrap http.ResponseWriter with enhanced logging and responses
- all: close out db connections where they've been missed

BUG FIXES

- internal/filetransfer: fix timezone issues in CutoffTime tests
- depositories: always set proper content-type in HTTP routes
- micro-deposits: grow mysql file_id column to store '*-micro-deposit-verify' IDs
- common: Fixed an issue in Amount.Equal()
- api,client: require x-user-id HTTP header in OpenAPI spec
- all: check sql Row.Scan errors
- transfers: expand window for EffectiveEntryDate comparison against created_at

cmd/server/main.go
// Copyright 2020 The Moov Authors
// Use of this source code is governed by an Apache License
// license that can be found in the LICENSE file.

package main

import (
	"context"
	"crypto/tls"
	"database/sql"
	"flag"
	"fmt"
	"net/http"
	"os"
	"os/signal"
	"path/filepath"
	"syscall"
	"time"

	"github.com/moov-io/base/admin"
	"github.com/moov-io/base/http/bind"
	"github.com/moov-io/paygate"
	"github.com/moov-io/paygate/internal/accounts"
	appcfg "github.com/moov-io/paygate/internal/config"
	"github.com/moov-io/paygate/internal/customers"
	"github.com/moov-io/paygate/internal/database"
	"github.com/moov-io/paygate/internal/depository"
	"github.com/moov-io/paygate/internal/depository/depadmin"
	"github.com/moov-io/paygate/internal/depository/verification/microdeposit"
	"github.com/moov-io/paygate/internal/events"
	"github.com/moov-io/paygate/internal/features"
	"github.com/moov-io/paygate/internal/fed"
	"github.com/moov-io/paygate/internal/filetransfer"
	ftadmin "github.com/moov-io/paygate/internal/filetransfer/admin"
	"github.com/moov-io/paygate/internal/filetransfer/config"
	"github.com/moov-io/paygate/internal/gateways"
	"github.com/moov-io/paygate/internal/model"
	"github.com/moov-io/paygate/internal/ofac"
	"github.com/moov-io/paygate/internal/originators"
	"github.com/moov-io/paygate/internal/receivers"
	"github.com/moov-io/paygate/internal/route"
	"github.com/moov-io/paygate/internal/secrets"
	"github.com/moov-io/paygate/internal/transfers"
	"github.com/moov-io/paygate/internal/transfers/limiter"
	"github.com/moov-io/paygate/internal/transfers/xferadmin"
	"github.com/moov-io/paygate/internal/util"

	"github.com/go-kit/kit/log"
	"github.com/gorilla/mux"
)

var (
	httpAddr  = flag.String("http.addr", bind.HTTP("paygate"), "HTTP listen address")
	adminAddr = flag.String("admin.addr", bind.Admin("paygate"), "Admin HTTP listen address")

	flagConfigFile = flag.String("config", "", "Filepath for config file to load")
	flagLogFormat  = flag.String("log.format", "", "Format for log lines (Options: json, plain")
)

func main() {
	flag.Parse()

	configFilepath := util.Or(os.Getenv("CONFIG_FILE"), *flagConfigFile)
	cfg, err := appcfg.LoadConfig(configFilepath, flagLogFormat)
	if err != nil {
		panic(fmt.Sprintf("failed to load config: %v", err))
	}
	cfg.Logger.Log("startup", fmt.Sprintf("Starting paygate server version %s", paygate.Version))

	ctx, cancelFunc := context.WithCancel(context.Background())
	defer cancelFunc()

	// migrate database
	db, err := database.New(ctx, cfg.Logger, database.Type())
	if err != nil {
		panic(fmt.Sprintf("error creating database: %v", err))
	}
	defer func() {
		if err := db.Close(); err != nil {
			cfg.Logger.Log("exit", err)
		}
	}()

	// Listen for application termination.
	errs := make(chan error)
	go func() {
		c := make(chan os.Signal, 1)
		signal.Notify(c, syscall.SIGINT, syscall.SIGTERM)
		errs <- fmt.Errorf("%s", <-c)
	}()

	// Spin up admin HTTP server and optionally override -admin.addr
	if v := os.Getenv("HTTP_ADMIN_BIND_ADDRESS"); v != "" {
		*adminAddr = v
	}
	adminServer := admin.NewServer(*adminAddr)
	adminServer.AddVersionHandler(paygate.Version) // Setup 'GET /version'
	go func() {
		cfg.Logger.Log("admin", fmt.Sprintf("listening on %s", adminServer.BindAddr()))
		if err := adminServer.Listen(); err != nil {
			err = fmt.Errorf("problem starting admin http: %v", err)
			cfg.Logger.Log("admin", err)
			errs <- err
		}
	}()
	defer adminServer.Shutdown()

	keeper, err := secrets.OpenSecretKeeper(ctx, "paygate-account-numbers", os.Getenv("CLOUD_PROVIDER"))
	if err != nil {
		panic(err)
	}
	stringKeeper := secrets.NewStringKeeper(keeper, 10*time.Second)

	// Setup repositories
	receiverRepo := receivers.NewReceiverRepo(cfg.Logger, db)
	defer receiverRepo.Close()

	depositoryRepo := depository.NewDepositoryRepo(cfg.Logger, db, stringKeeper)
	defer depositoryRepo.Close()

	// if err := internal.EncryptStoredAccountNumbers(cfg.Logger, depositoryRepo, stringKeeper); err != nil {
	// 	panic(err)
	// }

	eventRepo := events.NewRepo(cfg.Logger, db)
	defer eventRepo.Close()

	gatewaysRepo := gateways.NewRepo(cfg.Logger, db)
	defer gatewaysRepo.Close()

	originatorsRepo := originators.NewOriginatorRepo(cfg.Logger, db)
	defer originatorsRepo.Close()

	transferRepo := transfers.NewTransferRepo(cfg.Logger, db)
	defer transferRepo.Close()

	httpClient, err := route.TLSHttpClient(os.Getenv("HTTP_CLIENT_CAFILE"))
	if err != nil {
		panic(fmt.Sprintf("problem creating TLS ready *http.Client: %v", err))
	}

	// Create our various Client instances
	fedClient := setupFEDClient(cfg.Logger, os.Getenv("FED_ENDPOINT"), os.Getenv("FED_CALLS_DISABLED"), adminServer, httpClient)

	// Bring up our Accounts Client
	accountsClient := setupAccountsClient(cfg.Logger, adminServer, httpClient, os.Getenv("ACCOUNTS_ENDPOINT"), os.Getenv("ACCOUNTS_CALLS_DISABLED"))
	accountsCallsDisabled := accountsClient == nil

	customersClient := setupCustomersClient(cfg, adminServer, httpClient)
	customersCallsDisabled := customersClient == nil

	customerOFACRefresher := setupOFACRefresher(cfg, customersClient, db)
	if customerOFACRefresher != nil {
		defer customerOFACRefresher.Close()
	}

	features.AddRoutes(cfg.Logger, adminServer, accountsCallsDisabled, customersCallsDisabled)

	// Start our periodic file operations
	fileTransferRepo := config.NewRepository(configFilepath, db, os.Getenv("DATABASE_TYPE"))
	defer fileTransferRepo.Close()
	if err := config.ValidateTemplates(fileTransferRepo); err != nil {
		panic(fmt.Sprintf("ERROR: problem validating outbound filename templates: %v", err))
	}

	// micro-deposit repository
	microDepositRepo := microdeposit.NewRepository(cfg.Logger, db)

	// Register the depository admin routes
	depadmin.RegisterRoutes(cfg.Logger, adminServer, depositoryRepo)

	// Create HTTP handler
	handler := mux.NewRouter()
	receivers.AddReceiverRoutes(cfg.Logger, handler, customersClient, depositoryRepo, receiverRepo)
	events.AddRoutes(cfg.Logger, handler, eventRepo)
	gateways.AddRoutes(cfg.Logger, handler, gatewaysRepo)
	originators.AddOriginatorRoutes(cfg.Logger, handler, accountsClient, customersClient, depositoryRepo, originatorsRepo)
	route.AddPingRoute(cfg.Logger, handler)

	// Depository HTTP routes
	depositoryRouter := depository.NewRouter(cfg.Logger, fedClient, depositoryRepo, eventRepo, stringKeeper)
	depositoryRouter.RegisterRoutes(handler)

	// Gateway HTTP Routes
	gatewayRepo := gateways.NewRepo(cfg.Logger, db)
	gateways.AddRoutes(cfg.Logger, handler, gatewayRepo)

	// MicroDeposit HTTP routes
	attempter := microdeposit.NewAttemper(cfg.Logger, db, 5)
	odfiAccount := setupODFIAccount(accountsClient, stringKeeper)
	microDepositRouter := microdeposit.NewRouter(cfg.Logger, odfiAccount, attempter, accountsClient, depositoryRepo, eventRepo, gatewayRepo, microDepositRepo)
	microDepositRouter.RegisterRoutes(handler)

	// Transfer HTTP routes
	limits, err := limiter.Parse(limiter.OneDay(), limiter.SevenDay(), limiter.ThirtyDay())
	if err != nil {
		panic(fmt.Sprintf("ERROR parsing transfer limits: %v", err))
	}
	transferLimitChecker := limiter.New(cfg.Logger, db, limits)
	xferRouter := transfers.NewTransferRouter(cfg.Logger,
		depositoryRepo, eventRepo, gatewayRepo, receiverRepo, originatorsRepo, transferRepo,
		transferLimitChecker,
		accountsClient, customersClient,
	)
	xferRouter.RegisterRoutes(handler)
	xferadmin.RegisterRoutes(cfg.Logger, adminServer, transferRepo)

	// Async FileTransfer controller
	achStorageDir := setupACHStorageDir(cfg.Logger)
	fileTransferController, err := filetransfer.NewController(cfg, achStorageDir, fileTransferRepo, depositoryRepo, gatewaysRepo, microDepositRepo, originatorsRepo, receiverRepo, transferRepo, odfiAccount, accountsClient)
	if err != nil {
		panic(fmt.Sprintf("ERROR: creating ACH file transfer controller: %v", err))
	}
	shutdownFileTransferController := setupFileTransferController(cfg.Logger, fileTransferController, depositoryRepo, fileTransferRepo, microDepositRepo, transferRepo, adminServer)
	defer shutdownFileTransferController()

	// Check to see if our -http.addr flag has been overridden
	if v := os.Getenv("HTTP_BIND_ADDRESS"); v != "" {
		*httpAddr = v
	}
	// Create main HTTP server
	serve := &http.Server{
		Addr:    *httpAddr,
		Handler: handler,
		TLSConfig: &tls.Config{
			InsecureSkipVerify:       false,
			PreferServerCipherSuites: true,
			MinVersion:               tls.VersionTLS12,
		},
		ReadTimeout:  30 * time.Second,
		WriteTimeout: 30 * time.Second,
		IdleTimeout:  60 * time.Second,
	}
	shutdownServer := func() {
		if err := serve.Shutdown(context.TODO()); err != nil {
			cfg.Logger.Log("shutdown", err)
		}
	}
	defer shutdownServer()

	// Start main HTTP server
	go func() {
		if certFile, keyFile := os.Getenv("HTTPS_CERT_FILE"), os.Getenv("HTTPS_KEY_FILE"); certFile != "" && keyFile != "" {
			cfg.Logger.Log("startup", fmt.Sprintf("binding to %s for secure HTTP server", *httpAddr))
			if err := serve.ListenAndServeTLS(certFile, keyFile); err != nil {
				cfg.Logger.Log("exit", err)
			}
		} else {
			cfg.Logger.Log("startup", fmt.Sprintf("binding to %s for HTTP server", *httpAddr))
			if err := serve.ListenAndServe(); err != nil {
				cfg.Logger.Log("exit", err)
			}
		}
	}()

	if err := <-errs; err != nil {
		cfg.Logger.Log("exit", err)
	}
}

func setupAccountsClient(logger log.Logger, svc *admin.Server, httpClient *http.Client, endpoint, disabled string) accounts.Client {
	if util.Yes(disabled) {
		return nil
	}
	accountsClient := accounts.NewClient(logger, endpoint, httpClient)
	if accountsClient == nil {
		panic("no Accounts client created")
	}
	svc.AddLivenessCheck("accounts", accountsClient.Ping)
	return accountsClient
}

func setupCustomersClient(cfg *appcfg.Config, svc *admin.Server, httpClient *http.Client) customers.Client {
	if cfg.Customers.Disabled {
		return nil
	}
	client := customers.NewClient(cfg.Logger, cfg.Customers.Endpoint, httpClient)
	if client == nil {
		panic("no Customers client created")
	}
	svc.AddLivenessCheck("customers", client.Ping)
	return client
}

func setupOFACRefresher(cfg *appcfg.Config, client customers.Client, db *sql.DB) ofac.Refresher {
	refresher := ofac.NewRefresher(cfg, client, db)
	if refresher != nil {
		go func() {
			if err := refresher.Start(cfg.Customers.OFACRefreshEvery); err != nil {
				cfg.Logger.Log("customers", fmt.Errorf("problem with OFAC refresher: %v", err))
			}
		}()
	}
	return refresher
}

func setupFEDClient(logger log.Logger, endpoint string, disabled string, svc *admin.Server, httpClient *http.Client) fed.Client {
	if util.Yes(disabled) {
		return nil
	}
	client := fed.NewClient(logger, endpoint, httpClient)
	if client == nil {
		panic("no FED client created")
	}
	svc.AddLivenessCheck("fed", client.Ping)
	return client
}

func setupODFIAccount(accountsClient accounts.Client, keeper *secrets.StringKeeper) *microdeposit.ODFIAccount {
	odfiAccountType := model.Savings
	if v := os.Getenv("ODFI_ACCOUNT_TYPE"); v != "" {
		t := model.AccountType(v)
		if err := t.Validate(); err == nil {
			odfiAccountType = t
		}
	}

	accountNumber := util.Or(os.Getenv("ODFI_ACCOUNT_NUMBER"), "123")
	routingNumber := util.Or(os.Getenv("ODFI_ROUTING_NUMBER"), "121042882")

	return microdeposit.NewODFIAccount(accountsClient, accountNumber, routingNumber, odfiAccountType, keeper)
}

func setupACHStorageDir(logger log.Logger) string {
	dir := filepath.Dir(os.Getenv("ACH_FILE_STORAGE_DIR"))
	if dir == "." {
		dir = "./storage/"
	}
	if err := os.MkdirAll(dir, 0777); err != nil {
		panic(fmt.Sprintf("problem creating %s: %v", dir, err))
	}
	logger.Log("storage", fmt.Sprintf("using %s for storage directory", dir))
	return dir
}

func setupFileTransferController(
	logger log.Logger,
	controller *filetransfer.Controller,
	depRepo depository.Repository,
	fileTransferRepo config.Repository,
	microDepositRepo microdeposit.Repository,
	transferRepo transfers.Repository,
	svc *admin.Server,
) context.CancelFunc {
	ctx, cancelFileSync := context.WithCancel(context.Background())

	if controller == nil {
		return cancelFileSync
	}

	// setup buffered channels which only allow one concurrent operation
	flushIncoming := make(ftadmin.FlushChan, 1)
	flushOutgoing := make(ftadmin.FlushChan, 1)

	// start our controller's operations in an anon goroutine
	go controller.StartPeriodicFileOperations(ctx, flushIncoming, flushOutgoing)

	config.AddFileTransferConfigRoutes(logger, svc, fileTransferRepo)
	ftadmin.RegisterAdminRoutes(logger, svc, flushIncoming, flushOutgoing, func() ([]string, error) {
		return controller.GetMergedFilepaths()
	})

	return cancelFileSync
}
